certora/specs/Enabled.spec

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+import "DistinctIdentifiers.spec";
+
+function isInWithdrawQueueIsEnabled(uint256 i) returns bool {
+    if(i >= withdrawQueueLength()) return true;
+
+    MetaMorphoHarness.Id id = withdrawQueue(i);
+    bool enabled;
+    _, enabled, _ = config(id);
+
+    return enabled;
+}
+
+// Check that markets in the withdraw queue are enabled.
+invariant inWithdrawQueueIsEnabled(uint256 i)
+    isInWithdrawQueueIsEnabled(i)
+filtered {
+    f -> f.selector != sig:updateWithdrawQueue(uint256[]).selector
+}
+
+rule inWithdrawQueueIsEnabledPreservedUpdateWithdrawQueue(env e, uint256 i, uint256[] indexes) {
+    uint256 j;
+    require isInWithdrawQueueIsEnabled(indexes[i]);
+
+    // Safe require because it is a verified invariant.
+    require hasDistinctIdentifiers(indexes[i], j);
+
+    updateWithdrawQueue(e, indexes);
+
+    MetaMorphoHarness.Id id = withdrawQueue(i);
+    // Safe require because j is not otherwise constrained.
+    // The ghost variable deletedAt is useful to make sure that markets are not permuted and deleted at the same time in updateWithdrawQueue.
```

Comment:
it's technically impossible to do this right? So it's just a step to help the verifier realize this property and be able to prove the rule?

Reply:
Yes it's technically impossible because there are no duplicate Id in the withdraw queue. "Explaining" this to the prover is precisely why we introduce this `deletedAt` ghost variable. So yes it's a step to help the verifier (although my guess is that it would not be able to find this intermediary step by itself in a million years)